File: bizos-frontend/src/lib/api/food-vendor.ts
import { api } from './client';
import { withOfflineCache } from '@/lib/db/offlineCache';
import { FoodCredit, FoodCreditCreate, FoodVendorPayment } from '@/types/api';

export interface FoodVendorAnalytics {
  weekly_total: number;
  monthly_total: number;
  daily_average: number;
  total_outstanding: number;
  total_paid: number;
  total_credits: number;
  unpaid_count: number;
}

export interface FoodTrendPoint {
  date: string;
  total: number;
  count: number;
}

export interface VendorSpendingSummary {
  vendor_name: string;
  total_spent: number;
  total_meals: number;
  unpaid_amount: number;
}

export interface FoodMonthSummary {
  month: string;        // "YYYY-MM"
  total_spent: number;
  total_paid: number;
  total_credits: number;
  payment_count: number;
}

export const foodVendorApi = {
  credits: {
    list: (params?: { paid?: boolean }): Promise<FoodCredit[]> => {
      const qs = new URLSearchParams();
      if (params?.paid !== undefined) qs.set('paid', String(params.paid));
      return withOfflineCache(`food-credits-${params?.paid ?? 'all'}`, () =>
        api.get<FoodCredit[]>(`/food-vendor/credits?${qs}`));
    },
    listAll: (): Promise<FoodCredit[]> =>
      withOfflineCache('food-credits-all', () =>
        api.get<FoodCredit[]>('/food-vendor/credits')),
    create: (data: FoodCreditCreate) => api.post<FoodCredit>('/food-vendor/credits', data),
    update: (id: string, data: Partial<FoodCreditCreate>) =>
      api.patch<FoodCredit>(`/food-vendor/credits/${id}`, data),
    delete: (id: string) => api.delete<void>(`/food-vendor/credits/${id}`),
  },
  pay: (credit_ids: string[], vendor_name: string = 'Food Vendor') =>
    api.post<FoodVendorPayment>('/food-vendor/pay', { credit_ids, vendor_name }),
  payments: (): Promise<FoodVendorPayment[]> =>
    withOfflineCache('food-payments', () => api.get<FoodVendorPayment[]>('/food-vendor/payments')),
  analytics: (): Promise<FoodVendorAnalytics> =>
    withOfflineCache('food-analytics', () => api.get<FoodVendorAnalytics>('/food-vendor/analytics')),
  trend: (days = 30): Promise<FoodTrendPoint[]> =>
    withOfflineCache(`food-trend-${days}`, () => api.get<FoodTrendPoint[]>(`/food-vendor/trend?days=${days}`)),
  vendorBreakdown: (): Promise<VendorSpendingSummary[]> =>
    withOfflineCache('food-vendor-breakdown', () => api.get<VendorSpendingSummary[]>('/food-vendor/vendor-breakdown')),
  monthlySummary: (months = 6): Promise<FoodMonthSummary[]> =>
    withOfflineCache(`food-monthly-${months}`, () => api.get<FoodMonthSummary[]>(`/food-vendor/monthly-summary?months=${months}`)),
};

File: bizos-backend/schemas/food_vendor.py
from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import UUID4, BaseModel, Field


class FoodCreditCreate(BaseModel):
    vendor_name: str = Field(min_length=1, max_length=200)
    meal_description: Optional[str] = Field(None, max_length=500)
    amount: Decimal = Field(gt=Decimal("0"))
    purchase_date: Optional[date] = None


class FoodCreditOut(BaseModel):
    id: UUID4
    vendor_name: str
    meal_description: Optional[str]
    amount: Decimal
    purchase_date: date
    paid: bool
    paid_at: Optional[datetime]
    payment_batch_id: Optional[UUID4]
    created_at: datetime

    class Config:
        from_attributes = True


class FoodVendorPayRequest(BaseModel):
    credit_ids: List[UUID4] = Field(min_length=1)
    vendor_name: str = Field(min_length=1, max_length=200)


class FoodVendorPaymentOut(BaseModel):
    id: UUID4
    vendor_name: str
    amount_paid: Decimal
    paid_at: datetime
    note: Optional[str]

    class Config:
        from_attributes = True


class VendorOutstanding(BaseModel):
    vendor_name: str
    total_outstanding: Decimal
    credit_count: int


class FoodVendorAnalytics(BaseModel):
    weekly_total: Decimal
    monthly_total: Decimal
    daily_average: Decimal
    total_outstanding: Decimal
    total_paid: Decimal
    total_credits: int
    unpaid_count: int


class FoodTrendPoint(BaseModel):
    date: str
    total: Decimal
    count: int


class VendorSpendingSummary(BaseModel):
    vendor_name: str
    total_spent: Decimal
    total_meals: int
    unpaid_amount: Decimal


class FoodMonthSummary(BaseModel):
    month: str          # "YYYY-MM"
    total_spent: Decimal
    total_paid: Decimal
    total_credits: int
    payment_count: int
